TestingLowerBounds/KullbackLeibler.lean

```diff
@@ -599,51 +607,36 @@ lemma kl_fst_add_condKL [StandardBorelSpace β] [Nonempty β] {μ ν : Measure (
-of kernels. -/
-lemma kernel.absolutelyContinuous_compProd_iff [CountableOrCountablyGenerated β γ]
-    {κ₁ η₁ : kernel α β} {κ₂ η₂ : kernel (α × β) γ} [IsSFiniteKernel κ₁] [IsSFiniteKernel η₁]
-    [IsMarkovKernel κ₂] [IsMarkovKernel η₂] (a : α) :
-    (κ₁ ⊗ₖ κ₂) a ≪ (η₁ ⊗ₖ η₂) a ↔ κ₁ a ≪ η₁ a ∧ ∀ᵐ b ∂κ₁ a, κ₂ (a, b) ≪ η₂ (a, b) := by
-  simp_rw [kernel.compProd_apply_eq_compProd_snd',
-    kernel.Measure.absolutelyContinuous_compProd_iff, kernel.snd'_apply]
-
-lemma kernel.integrable_llr_compProd_iff [CountableOrCountablyGenerated β γ] {κ₁ η₁ : kernel α β}
-    {κ₂ η₂ : kernel (α × β) γ} [IsFiniteKernel κ₁] [IsFiniteKernel η₁] [IsMarkovKernel κ₂]
-    [IsMarkovKernel η₂] (a : α) (h_ac : (κ₁ ⊗ₖ κ₂) a ≪ (η₁ ⊗ₖ η₂) a) :
+/-TODO: this is just a thin wrapper around kernel.integrable_llr_compProd_iff, so that that lemma
+could be put in an outside file. But I have realised that the choice of having 2 instead of 2' as
+the hp of choice about integrability here may be a bad one, because in cases like this one
+it does not allow to move stuff outside this file, as it relies on the definition of kl.
+Moreover in general it is the opposite choice to what is done in fDiv, and in fDiv the other choice
+is much more convenient, because it allows to disregard the singular part inside the definition of
+fDiv when talking about integrability. So I think it may be better to reverse this choice here,
+changing the lemmas like condKL_ne_top_iff from 2 to 2'-/
+lemma kernel.integrable_llr_compProd_iff' [CountableOrCountablyGenerated (α × β) γ]
+    {κ₁ η₁ : kernel α β} {κ₂ η₂ : kernel (α × β) γ} [IsFiniteKernel κ₁] [IsFiniteKernel η₁]
+    [IsMarkovKernel κ₂] [IsMarkovKernel η₂] (a : α) (h_ac : (κ₁ ⊗ₖ κ₂) a ≪ (η₁ ⊗ₖ η₂) a) :
     Integrable (llr ((κ₁ ⊗ₖ κ₂) a) ((η₁ ⊗ₖ η₂) a)) ((κ₁ ⊗ₖ κ₂) a)
       ↔ Integrable (llr (κ₁ a) (η₁ a)) (κ₁ a)
         ∧ Integrable (fun b ↦ (kl (κ₂ (a, b)) (η₂ (a, b))).toReal) (κ₁ a)
         ∧ ∀ᵐ b ∂κ₁ a, Integrable (llr (κ₂ (a, b)) (η₂ (a, b))) (κ₂ (a, b)) := by
+  convert kernel.integrable_llr_compProd_iff a h_ac using 3
+  simp_rw [← kernel.snd'_apply]
   have h_ac' := kernel.absolutelyContinuous_compProd_iff a |>.mp h_ac |>.2
-  simp_rw [kernel.compProd_apply_eq_compProd_snd'] at h_ac
-  simp_rw [kernel.compProd_apply_eq_compProd_snd',
-    ProbabilityTheory.integrable_llr_compProd_iff h_ac, kernel.snd'_apply]
-  by_cases h_int₁ : Integrable (llr (κ₁ a) (η₁ a)) (κ₁ a)
-  swap; tauto
-  by_cases h_int₂ : ∀ᵐ b ∂κ₁ a, Integrable (llr (κ₂ (a, b)) (η₂ (a, b))) (κ₂ (a, b))
-  swap; tauto
-  simp only [h_int₁, true_and, h_int₂, and_true]
-  apply integrable_congr
-  filter_upwards [h_ac'] with b hb_ac
-  rw [kl_toReal_of_ac hb_ac]
+  exact integrable_kl_iff h_ac'
 
-lemma kl_compProd_kernel_of_ae_ac_of_ae_integrable [CountableOrCountablyGenerated β γ]
+lemma kl_compProd_kernel_of_ae_ac_of_ae_integrable [CountableOrCountablyGenerated (α × β) γ]
```

Comment:
This changed from `[CountableOrCountablyGenerated β γ]` to `[CountableOrCountablyGenerated (α × β) γ]`. Can you comment on that?